Update database access docs

r0mant opened the pull request:
A few updates/enhancements to the database access preview docs to make them more complete and useful for general audience:

* Add a section on configuring a self-hosted Postgres server.
* Add a section on starting database service w/o configuration file and generating dynamic tokens.
* Add more information around CLI UX.
* Add a section about RBAC.
* Fix RFD link.

r0mant committed:
Update database access docs

klizhentas commented:
@r0mant run `make docs-fix-whitespace` to fix the CI/CD problem

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
+
+!!! note
+    This section explains how to configure a PostgreSQL-flavored instance of
+    AWS RDS or Aurora database to work with Teleport database access. For
```
```suggestion
    AWS RDS or Aurora database to work with Teleport Database Access. For
```
for consistency

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
-database to it via Teleport database proxy.
+!!! note
+    This section explains how to configure an on-premise instance of PostgreSQL
+    to work with Teleport Database access. For information about configuring
```
```suggestion
    to work with Teleport Database Access. For information about configuring
```
for consistency

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -149,34 +208,70 @@ ssh_service:
-Create configuration file for a database service:
+Note that the `--auth-server` flag must point to cluster's proxy endpoint
+because database service always connects back to the cluster over reverse
+tunnel framework.
```
```suggestion
tunnel.
```
b/c implementation detail

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -149,34 +208,70 @@ ssh_service:
+
+### Start Database Service with Config File
+
+Below is the example of a database service configuration file that proxies
```
```suggestion
Below is an example of a database service configuration file that proxies
```

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -149,34 +208,70 @@ ssh_service:
 db_service:
   enabled: "yes"
+  # This section contains definitions of all databases proxied by this
+  # service, can contain multiple.
```
```suggestion
  # service, can contain multiple items.
```

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,122 @@ proxy_service:
-Start the service:
+!!! tip
+    Single Teleport process can run multiple different services, for example
+    multiple database access proxies as well as run other services such as
```
```suggestion
    multiple database access proxies as well as an SSH service or an application access proxy.
```

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,122 @@ proxy_service:
+!!! tip
+    You can be logged into multiple databases simultaneously.
+
+You can optionally also specify default database name and user to use when
```
```suggestion
You can optionally also specify the database name and the user to use when
```

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,122 @@ proxy_service:
+## RBAC
+
+Teleport's "role" resource provides the following instruments for restricting
+the database access:
```
```suggestion
database access:
```

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,122 @@ proxy_service:
+    # service.
+    db_labels:
+      environment: ["dev", "stage"]
+    # Database names this role has connect to. Note: this is not the same as
```
```suggestion
    # Database names this role will be able to connect to. Note: this is not the same as
```

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
-!!! info
-    Teleport Database Access is under active development.
-    The alpha release will include support for PostgreSQL, including Amazon RDS and Aurora.
+Teleport Database Access is under active development. The alpha release will
```
```suggestion
Teleport Database Access is under active development. The first version will 
```

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
+## Configure PostgreSQL
 
-Let's setup a sample Teleport Database Access deployment.
+### On-Prem PostgreSQL
```
```suggestion
### Self-Hosted PostgreSQL
```

We are using self-hosted vs on-prem. As on-prem can be a limited term vs 'self-hosted' which includes people running PostgreSQL themselves on a cloud provider.

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
-We will create a Teleport cluster and connect a PostgreSQL-flavored AWS Aurora
-database to it via Teleport database proxy.
+!!! note
+    This section explains how to configure an on-premise instance of PostgreSQL
```
```suggestion
    This section explains how to configure a self-hosted instance of PostgreSQL
```

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
 
-Teleport Database Access uses IAM authentication with AWS RDS and Aurora
-databases which can be enabled with the following steps.
+Teleport uses mutual TLS authentication so PostgreSQL server must be configured
```
I would expand this section / introduction. Maybe something like...

> Teleport uses mutual TLS to authentication to access to PostgreSQL instances. The PostgreSQL instance much be configured

benarent commented on docs/5.0/preview/database-access.md:
````diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
+on the auth server.
+
+```sh
+$ tctl auth sign --format=db --host=db.example.com --out=server --ttl=8760h
````
Added a comment for lazy people not reading below. Do we have any thoughts / recommendations for best practices for shorter lived certs / automation for getting new certs. 

```suggestion
# Export Teleport's certificate authority for host db.example.com with a one year validity period. 
$ tctl auth sign --format=db --host=db.example.com --out=server --ttl=8760h
```

benarent replied:
It would be good to expand on our best practices for certificate automation. 

Does Postgres have any best practices for dealing with rotating certs? e.g. Does `postgresql.conf` cache these credentials / What's the process for updating certs on production systems.

r0mant replied:
I don't think Postgres supports any sort of automation for certificates rotation, so it will likely require database restart.

For now, I've just added a note that it has to be considered when rotating host CA, we might need to expand on this topic more in one of future iterations.

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
+    AWS RDS or Aurora database to work with Teleport Database Access. For
+    information about configuring an on-prem PostgreSQL see the [section above](#on-prem-postgresql).
+
+Teleport Database Access for AWS RDS and Aurora uses IAM authentication which
```
We should add a note below about how to obtain the ARN for the IAM policy.

r0mant replied:
Done.

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -110,7 +169,7 @@ GRANT rds_iam TO alice;
-### Install Teleport
+## Configure Teleport
 
 First, head over to the Teleport [downloads page](https://gravitational.com/teleport/download/)
```
```suggestion
First, head over to the Teleport [downloads page](https://goteleport.com/teleport/download/)
```

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -121,7 +180,7 @@ and download the latest version of Teleport.
 
 Follow the installation [instructions](https://gravitational.com/teleport/docs/installation/).
```
```suggestion
Follow the installation [instructions](installation.md).
```

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -149,34 +208,70 @@ ssh_service:
+   --roles=database \
+   --token=cbdeeab9-6f88-436d-a673-44d14bd86bb7 \
+   --auth-server=teleport.example.com:3080 \
+   --db-name=test \
```
Does this need to match the name in the exported Teleport CA?

r0mant replied:
No, this is just the service name used for identification purposes and in RBAC labels.

benarent commented on docs/5.0/preview/database-access.md:
````diff
@@ -149,34 +208,70 @@ ssh_service:
+```
+
+Note that the `--auth-server` flag must point to cluster's proxy endpoint
+because database service always connects back to the cluster over reverse
````
```suggestion
because database service always connects back to the cluster over a reverse
```

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -149,34 +208,70 @@ ssh_service:
     uri: "postgres-aurora-instance-1.xxx.us-east-1.rds.amazonaws.com:5432"
-    # AWS specific configuration.
+    # AWS specific configuration, only required for RDS and Aurora.
     aws:
       # Region the database is deployed in.
       region: us-east-1
```
Do we really need this extra setting if we already know the location of the database via the uri? 

postgres-aurora-instance-1.xxx.**us-east-1**.rds.amazonaws.com:5432

r0mant replied:
I thought about this. The thing is, I'm also using this field to tell that this is actually RDS/Aurora instance (to use IAM auth). I guess we could also just try to determine this from the URI as well, but at that point I feel like it would be trying to be too smart and deduce too much which may lead to confusing behaviors (i.e. could you have a different URI for RDS, like a cname? I don't know)... so I usually prefer to be explicit. It is pretty tempting though to get rid of an extra setting.

benarent commented on docs/5.0/preview/database-access.md:
````diff
@@ -191,30 +286,140 @@ proxy_service:
 ```
 
-### Connect
+Note that you will only be able to see databases your role has access to. See
````
```suggestion
Note you will only be able to see databases your role has access to. See
```

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,140 @@ proxy_service:
+!!! tip
+    You can be logged into multiple databases simultaneously.
+
+You can optionally also specify the database name and the user to use by
```
```suggestion
You can optionally specify the database name and the user to use by
```

benarent commented on docs/5.0/preview/database-access.md:
````diff
@@ -191,30 +286,140 @@ proxy_service:
+default when connecting to the database instance:
+
+```sh
+$ tsh db login --db-user=postgres --db-name=postgres aurora
````
Updated to match example. 
```suggestion
$ tsh db login --db-user=postgres --db-name=aurora
```

r0mant replied:
The current version is correct as it is actually. We're connecting to the database "postgres" within the database server "aurora".

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,140 @@ proxy_service:
+in the [connection service file](https://www.postgresql.org/docs/current/libpq-pgservice.html)
+with the name of `<cluster-name>-<database-service-name>`.
+
+Suppose the cluster name is "root", then you can connect to the database using
```
This is a little confusing. Is this 'root' as in Teleport Root Cluster? Or something else?

r0mant replied:
In this case "root" is the cluster name.

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,140 @@ proxy_service:
+    # within a particular database instance.
+    db_names: ["main", "metrics", "postgres"]
+    # Database users this role can connect as.
+    db_users: ["alice", "bob"]
```
It would be good to add add a deny example, and possibly highlight more advanced features such as wildcards. 

e.g. 

```deny:
      db_name: ["prod*"]

r0mant replied:
Added below.

benarent commented on docs/5.0/preview/database-access.md:
````diff
@@ -191,30 +286,140 @@ proxy_service:
+```yaml
+spec:
+  allow:
+    db_names: ["{% raw %}{{internal.db_names}}{% endraw %}", "{% raw %}{{external.xxx}}{% endraw %}"]
````
Side note: We should explain the difference between `{{internal` vs `{{external` does.

webvictim replied:
We do this elsewhere in our RBAC docs, no? We should just link to that section rather than copy/pasting here.

r0mant replied:
Yeah, there's a link to general RBAC docs right below this.

benarent commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
+
+* `--format=db`: instructs the command to produce secrets in the format suitable
+  for configuring a database server.
+* `--host=db.example.com`: server name to encode in the certificate, should
```
If using AWS does this also need to map to any AWS RDS name/setting?

r0mant replied:
No, generating certificates is only required for self-hosted instances. RDS users don't need to do this at all.

webvictim commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
-Teleport Database Access uses IAM authentication with AWS RDS and Aurora
-databases which can be enabled with the following steps.
+Teleport uses mutual TLS authentication so PostgreSQL server must be configured
+with Teleport's certificate authority and certificate/key pair that Teleport
```
```suggestion
with Teleport's certificate authority and a certificate/key pair that Teleport
```

webvictim commented on docs/5.0/preview/database-access.md:
```diff
@@ -44,23 +44,82 @@ The following diagram shows an example Database Access setup:
+!!! note
+    This section explains how to configure a PostgreSQL-flavored instance of
+    AWS RDS or Aurora database to work with Teleport Database Access. For
+    information about configuring an on-prem PostgreSQL see the [section above](#on-prem-postgresql).
```
```suggestion
    information about configuring self-hosted PostgreSQL see the [section above](#self-hosted-postgresql).
```

webvictim commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,140 @@ proxy_service:
 
-Start the service:
+!!! tip
+    Single Teleport process can run multiple different services, for example
+    multiple database access proxies as well as run other services such an
+    SSH service or an application access proxy.
```
```suggestion
    A single Teleport process can run multiple different services, for example
    multiple database access proxies as well as running other services such an
    SSH service or an application access proxy.
```

webvictim commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,140 @@ proxy_service:
+
+When setting up Teleport database service with AWS RDS or Aurora, it must have
+an IAM role allowing it to connect to that particular database instance. An
+example of such policy is shown in the [AWS RDS/Aurora](#aws-rdsaurora-postgresql)
```
```suggestion
example of such a policy is shown in the [AWS RDS/Aurora](#aws-rdsaurora-postgresql)
```

webvictim commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,140 @@ proxy_service:
+section above. See [Creating and using an IAM policy for IAM database access](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.IAMPolicy.html)
+in AWS documentation.
+
+Teleport database service uses default credential provider chain to find AWS
```
```suggestion
Teleport database service uses the default credential provider chain to find AWS
```

webvictim commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +286,140 @@ proxy_service:
+  Cancel requests issued by the PostgreSQL clients connected to Teleport proxy
+  won't be passed to the database server.
+* Any of the [authentication methods](https://www.postgresql.org/docs/current/auth-methods.html)
+  except for the client certificate authentication.
```
```suggestion
  except for client certificate authentication.
```

webvictim commented on docs/5.0/preview/database-access.md:
```diff
@@ -223,10 +428,9 @@ $ psql "service=postgres-aurora user=<db-user> database=<db-name>"
 ## RFD
 
-Please refer to the [RFD document](https://github.com/gravitational/teleport/blob/roman/rfd/dba/rfd/0011-database-access.md)
+Please refer to the [RFD document](https://github.com/gravitational/teleport/blob/master/rfd/0011-database-access.md)
```
We should link to the commit hash in case it gets renamed/moved.

```suggestion
Please refer to the [RFD document](https://github.com/gravitational/teleport/blob/024fd582e62560ffc925cd4b4d42c156043c041b/rfd/0011-database-access.md)
```

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -135,7 +216,7 @@ auth_service:
   listen_addr: 0.0.0.0:3025
   tokens:
-  - proxy,node,database:qwe123
+  - proxy,node,database:cbdeeab9-6f88-436d-a673-44d14bd86bb7
```
nit: I know it must have been suggested to use token that resembled a real token, but since this needs to be matched in multiple commands and yet be easier to compare/remember when skimming through the docs - how about using something more memorable like `00000000-0000-0000-0000-000000000001`?

r0mant replied:
I think using a real uuid sends a stronger signal to readers that using a secure token is advised.

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -149,34 +230,70 @@ ssh_service:
+    --type=database \
+    --db-name=test \
+    --db-protocol=postgres \
+    --db-uri=db.example.com:5432
```
nit: can we use an example output to hint at what a dynamic token could look like?

r0mant replied:
Not sure it'd add a lot of value to the reader to be honest, outputs also have a tendency to change / become obsolete.

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -149,34 +230,70 @@ ssh_service:
+  auth_token: cbdeeab9-6f88-436d-a673-44d14bd86bb7
+  # Proxy address to connect to. Note that it has to be the proxy address
+  # because database service always connects to the cluster over reverse
+  # tunnel framework.
```
```suggestion
  # tunnel.
```

a-palchikov commented on docs/5.0/preview/database-access.md:
```diff
@@ -191,30 +308,162 @@ proxy_service:
+
+### Which database protocols does Teleport Database Access support?
+
+Teleport Database Access Preview supports PostgreSQL protocol. MySQL protocol
```
```suggestion
Teleport Database Access Preview only supports the PostgreSQL protocol. MySQL protocol
```

klizhentas commented:
@r0mant let's rebase, fix conflicts and merge

r0mant commented:
@klizhentas I've addressed comments, rebased and fixed conflicts, need at least 1 approval in order to avoid merge policy override. cc @benarent @webvictim
